conu/backend/nspawn/image.py

```diff
@@ -0,0 +1,351 @@
+            logger.debug("pull policy set to 'if_not_present' and image is not present, "
+                         "pulling the image")
+            self.pull()
+        elif self.pull_policy == ImagePullPolicy.NEVER:
+            logger.debug("pull policy set to 'never'")
+
+    def __repr__(self):
+        return "DockerImage(repository=%s, tag=%s)" % (self.name, self.tag)
+
+    def __str__(self):
+        return self.get_full_name()
+
+    def get_full_name(self):
+        """
+        Provide full, complete image name
+
+        :return: str
+        """
+        return "%s:%s" % (self.name, self.tag)
+
+    def get_id(self):
+        """
+        get unique identifier of this image
+
+        :return: str
+        """
+        if self._id is None:
+            self._id = "{}:{}".format(hashlib.sha512(self.name.encode("ascii")).hexdigest(), self.tag)
+        return self._id
+
+    def is_present(self):
+        return os.path.exists(self.image_location)
+
+    def pull(self):
+        """
+        Pull this image from URL. Raises an exception if the image is not found in
+        the registry.
+
+        :return: None
+        """
+        if self.use_as_image:
+            try:
+                logger.debug("Try to pull URL: {} -> {}".format(self.name, self.image_location))
+                if self.name.endswith(".xz"):
+                    logger.info("uncompressing image, can take a while")
+                    # TODO: improve this part
+                    decomp = lzma.LZMADecompressor()
+                    data = urllib.urlopen(self.name)
+                    with open(self.image_location, mode="wb") as output:
+                        while True:
+                            comp_data = data.read()
+                            if not comp_data:
+                                data.close()
+                                break
+                            uncomp_data = decomp.decompress(comp_data)
+                            output.write(uncomp_data)
+                else:
+                    urllib.urlretrieve(self.name, self.image_location)
+            except ValueError as error:
+                    raise ConuException("There was an error while pulling the image %s: %s",
+                                       self.name, error)
+        else:
+            logger.debug("Skip pull image, repository is local FS location")
+
+    def create_snapshot(self, destination, tag):
+        """
+        returns Image object with copyied files from base image
+
+        :param destination: directory where to crete copy
+        :return: Image
+        """
+        logger.debug("Create Snapshot: %s -> %s" % (self.image_location, destination))
+        # copytree somethimes fails, it is not reliable in case of copy of system
+        # shutil.copytree(self.location, destination)
+        # cp will do better work
+        run_cmd(["cp", "-rf", self.image_location, destination])
```

Comment:
setting up ostree or overlay would be much better, this seems good for start though